File: src/apps/daemon/internal/networking/transport/replay_resistant_tunnel.go
package transport

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"io"

	"golang.org/x/crypto/hkdf"
)

type ReplayResistantConfig struct {
	PSK              []byte
	MaxSkewSecs      int64
	MaxTrackedNonces int
}

func DefaultReplayResistantConfig() ReplayResistantConfig {
	return ReplayResistantConfig{
		PSK:              []byte("luminet-default-replay-psk-2026"),
		MaxSkewSecs:      60,
		MaxTrackedNonces: 4096,
	}
}

type ReplayResistantTunnelSession struct {
	config        ReplayResistantConfig
	sessionKey    []byte
	lastRemoteSeq uint64
	seenNonces    map[[12]byte]struct{}
}

func NewReplayResistantTunnelSession(config ReplayResistantConfig, clientRandom, serverRandom []byte) *ReplayResistantTunnelSession {
	hk := hkdf.New(sha256.New, config.PSK, clientRandom, serverRandom)
	sessionKey := make([]byte, 32)
	if _, err := io.ReadFull(hk, sessionKey); err != nil {
		// Fallback deterministic expand
		h := sha256.Sum256(append(config.PSK, append(clientRandom, serverRandom...)...))
		copy(sessionKey, h[:])
	}

	return &ReplayResistantTunnelSession{
		config:        config,
		sessionKey:    sessionKey,
		lastRemoteSeq: 0,
		seenNonces:    make(map[[12]byte]struct{}),
	}
}

func (s *ReplayResistantTunnelSession) SealPacket(sequence uint64, timestamp int64, payload []byte) []byte {
	packet := make([]byte, 8+8+12+len(payload))
	binary.BigEndian.PutUint64(packet[0:8], sequence)
	binary.BigEndian.PutUint64(packet[8:16], uint64(timestamp))

	var nonce [12]byte
	seqBytes := packet[0:8]
	tsBytes := packet[8:16]
	for i := 0; i < 8; i++ {
		nonce[i] = seqBytes[i] ^ s.sessionKey[i]
	}
	for i := 0; i < 4; i++ {
		nonce[8+i] = tsBytes[i] ^ s.sessionKey[8+i]
	}
	copy(packet[16:28], nonce[:])

	for i, b := range payload {
		k := s.sessionKey[(i+int(nonce[i%12]))%len(s.sessionKey)]
		packet[28+i] = b ^ k
	}

	mac := hmac.New(sha256.New, s.sessionKey)
	mac.Write(packet)
	tag := mac.Sum(nil)
	packet = append(packet, tag[:16]...)

	return packet
}

func (s *ReplayResistantTunnelSession) OpenPacket(packet []byte, currentTime int64) (uint64, int64, []byte, error) {
	if len(packet) < 8+8+12+16 {
		return 0, 0, nil, errors.New("packet smaller than replay envelope minimum")
	}

	sequence := binary.BigEndian.Uint64(packet[0:8])
	timestamp := int64(binary.BigEndian.Uint64(packet[8:16]))

	diff := currentTime - timestamp
	if diff < 0 {
		diff = -diff
	}
	if diff > s.config.MaxSkewSecs {
		return 0, 0, nil, fmt.Errorf("timestamp skew too large: %d s", diff)
	}

	var nonce [12]byte
	copy(nonce[:], packet[16:28])

	if _, exists := s.seenNonces[nonce]; exists {
		return 0, 0, nil, errors.New("replay detected: duplicate nonce")
	}

	bodyEnd := len(packet) - 16
	providedTag := packet[bodyEnd:]

	mac := hmac.New(sha256.New, s.sessionKey)
	mac.Write(packet[:bodyEnd])
	expectedTag := mac.Sum(nil)[:16]
	if !hmac.Equal(providedTag, expectedTag) {
		return 0, 0, nil, errors.New("integrity verification failed")
	}

	if sequence <= s.lastRemoteSeq && s.lastRemoteSeq > 0 {
		return 0, 0, nil, fmt.Errorf("out of order sequence: got %d, last %d", sequence, s.lastRemoteSeq)
	}

	encPayload := packet[28:bodyEnd]
	decrypted := make([]byte, len(encPayload))
	for i, b := range encPayload {
		k := s.sessionKey[(i+int(nonce[i%12]))%len(s.sessionKey)]
		decrypted[i] = b ^ k
	}

	if len(s.seenNonces) >= s.config.MaxTrackedNonces {
		s.seenNonces = make(map[[12]byte]struct{})
	}
	s.seenNonces[nonce] = struct{}{}
	s.lastRemoteSeq = sequence

	return sequence, timestamp, decrypted, nil
}

File: src/packages/lumicore/src/transport/replay_resistant_tunnel.rs
//! # Replay-Resistant Tunnel Session
//!
//! Authenticated proxy tunnel session protocol preventing packet replay attacks
//! through sliding nonce tracking, monotonic sequence validation, and HKDF session key derivation.

use hkdf::Hkdf;
use sha2::Sha256;
use std::collections::HashSet;

#[derive(Debug, Clone)]
pub struct ReplayResistantConfig {
    pub psk: Vec<u8>,
    pub max_skew_secs: u64,
    pub max_tracked_nonces: usize,
}

impl Default for ReplayResistantConfig {
    fn default() -> Self {
        Self {
            psk: b"luminet-default-replay-psk-2026".to_vec(),
            max_skew_secs: 60,
            max_tracked_nonces: 4096,
        }
    }
}

pub struct ReplayResistantTunnelSession {
    config: ReplayResistantConfig,
    session_key: Vec<u8>,
    last_remote_seq: u64,
    seen_nonces: HashSet<[u8; 12]>,
}

impl ReplayResistantTunnelSession {
    pub fn new(config: ReplayResistantConfig, client_random: &[u8], server_random: &[u8]) -> Self {
        let hk = Hkdf::<Sha256>::new(Some(client_random), &config.psk);
        let mut session_key = vec![0u8; 32];
        hk.expand(server_random, &mut session_key)
            .expect("32 bytes is valid length for HKDF-SHA256");

        Self {
            config,
            session_key,
            last_remote_seq: 0,
            seen_nonces: HashSet::new(),
        }
    }

    pub fn seal_packet(&self, sequence: u64, timestamp: u64, payload: &[u8]) -> Vec<u8> {
        let mut packet = Vec::with_capacity(8 + 8 + 12 + payload.len() + 16);
        packet.extend_from_slice(&sequence.to_be_bytes());
        packet.extend_from_slice(&timestamp.to_be_bytes());

        // Derive deterministic nonce from sequence, timestamp, and session_key
        let mut nonce = [0u8; 12];
        let seq_bytes = sequence.to_be_bytes();
        let ts_bytes = timestamp.to_be_bytes();
        for i in 0..8 {
            nonce[i] = seq_bytes[i] ^ self.session_key[i];
        }
        for i in 0..4 {
            nonce[8 + i] = ts_bytes[i] ^ self.session_key[8 + i];
        }
        packet.extend_from_slice(&nonce);

        // XOR payload with session stream
        for (i, byte) in payload.iter().enumerate() {
            let k = self.session_key[(i + nonce[i % 12] as usize) % self.session_key.len()];
            packet.push(byte ^ k);
        }

        // 16-byte MAC tag over packet
        let hk = Hkdf::<Sha256>::new(Some(&nonce), &self.session_key);
        let mut tag = [0u8; 16];
        hk.expand(&packet, &mut tag).expect("16 bytes valid");
        packet.extend_from_slice(&tag);

        packet
    }

    pub fn open_packet(
        &mut self,
        packet: &[u8],
        current_time: u64,
    ) -> Result<(u64, u64, Vec<u8>), String> {
        if packet.len() < 8 + 8 + 12 + 16 {
            return Err("Packet smaller than replay envelope minimum".to_string());
        }

        let seq_bytes: [u8; 8] = packet[0..8].try_into().unwrap();
        let sequence = u64::from_be_bytes(seq_bytes);

        let ts_bytes: [u8; 8] = packet[8..16].try_into().unwrap();
        let timestamp = u64::from_be_bytes(ts_bytes);

        // Validate timestamp skew
        let diff = if current_time >= timestamp {
            current_time - timestamp
        } else {
            timestamp - current_time
        };
        if diff > self.config.max_skew_secs {
            return Err(format!("Timestamp skew too large: {}s", diff));
        }

        let mut nonce = [0u8; 12];
        nonce.copy_from_slice(&packet[16..28]);

        // Check replay via nonce cache
        if self.seen_nonces.contains(&nonce) {
            return Err("Replay detected: duplicate nonce".to_string());
        }

        // Verify MAC
        let body_end = packet.len() - 16;
        let provided_tag = &packet[body_end..];
        let hk = Hkdf::<Sha256>::new(Some(&nonce), &self.session_key);
        let mut expected_tag = [0u8; 16];
        hk.expand(&packet[..body_end], &mut expected_tag).expect("16 bytes valid");
        if provided_tag != expected_tag {
            return Err("Integrity verification failed".to_string());
        }

        // Sequence monotonicity check
        if sequence <= self.last_remote_seq && self.last_remote_seq > 0 {
            return Err(format!(
                "Out of order / replayed sequence: got {}, last {}",
                sequence, self.last_remote_seq
            ));
        }

        // Decrypt payload
        let enc_payload = &packet[28..body_end];
        let mut decrypted = Vec::with_capacity(enc_payload.len());
        for (i, byte) in enc_payload.iter().enumerate() {
            let k = self.session_key[(i + nonce[i % 12] as usize) % self.session_key.len()];
            decrypted.push(byte ^ k);
        }

        // Commit nonce and sequence
        if self.seen_nonces.len() >= self.config.max_tracked_nonces {
            self.seen_nonces.clear();
        }
        self.seen_nonces.insert(nonce);
        self.last_remote_seq = sequence;

        Ok((sequence, timestamp, decrypted))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_seal_and_open_success() {
        let config = ReplayResistantConfig::default();
        let client_rand = b"client_nonce_123456";
        let server_rand = b"server_nonce_abcdef";

        let sender = ReplayResistantTunnelSession::new(config.clone(), client_rand, server_rand);
        let mut receiver = ReplayResistantTunnelSession::new(config, client_rand, server_rand);

        let now = 1700000000;
        let sealed = sender.seal_packet(1, now, b"authenticated message");

        let (seq, ts, payload) = receiver.open_packet(&sealed, now + 2).unwrap();
        assert_eq!(seq, 1);
        assert_eq!(ts, now);
        assert_eq!(payload, b"authenticated message");
    }

    #[test]
    fn test_replay_rejection() {
        let config = ReplayResistantConfig::default();
        let client_rand = b"client_nonce_123456";
        let server_rand = b"server_nonce_abcdef";

        let sender = ReplayResistantTunnelSession::new(config.clone(), client_rand, server_rand);
        let mut receiver = ReplayResistantTunnelSession::new(config, client_rand, server_rand);

        let now = 1700000000;
        let sealed = sender.seal_packet(1, now, b"packet one");

        assert!(receiver.open_packet(&sealed, now).is_ok());
        // Replaying same packet should fail
        let replay_err = receiver.open_packet(&sealed, now);
        assert!(replay_err.is_err());
    }

    #[test]
    fn test_skew_timeout_rejection() {
        let config = ReplayResistantConfig::default();
        let client_rand = b"client_nonce_123456";
        let server_rand = b"server_nonce_abcdef";

        let sender = ReplayResistantTunnelSession::new(config.clone(), client_rand, server_rand);
        let mut receiver = ReplayResistantTunnelSession::new(config, client_rand, server_rand);

        let now = 1700000000;
        let sealed = sender.seal_packet(1, now, b"packet one");

        // 120 seconds later (max skew is 60)
        let res = receiver.open_packet(&sealed, now + 120);
        assert!(res.is_err());
    }
}
